wen-restart/src/wen_restart.rs

```diff
@@ -126,27 +509,566 @@ mod tests {
+                &last_vote_hash,
+                &keypairs.node_keypair,
+                now,
+            );
+            expected_messages.insert(
+                node_pubkey.to_string(),
+                LastVotedForkSlotsRecord {
+                    last_voted_fork_slots: expected_slots_to_repair.clone(),
+                    last_vote_bankhash: last_vote_hash.to_string(),
+                    shred_version: SHRED_VERSION as u32,
+                    wallclock: now,
+                },
+            );
+        }
+
+        // Simulating successful repair of missing blocks.
+        insert_and_freeze_slots(test_state.bank_forks.clone(), expected_slots_to_repair);
+
+        let _ = wen_restart_thread_handle.join();
+        let progress = read_wen_restart_records(&test_state.wen_restart_proto_path).unwrap();
+        let progress_start_time = progress
+            .my_last_voted_fork_slots
+            .as_ref()
+            .unwrap()
+            .wallclock;
         assert_eq!(
             progress,
             WenRestartProgress {
-                state: RestartState::Init.into(),
-                my_last_voted_fork_slots: Some(MyLastVotedForkSlots {
-                    last_vote_slot,
+                state: RestartState::Done.into(),
+                my_last_voted_fork_slots: Some(LastVotedForkSlotsRecord {
+                    last_voted_fork_slots: test_state.last_voted_fork_slots,
+                    last_vote_bankhash: last_vote_bankhash.to_string(),
+                    shred_version: SHRED_VERSION as u32,
+                    wallclock: progress_start_time,
+                }),
+                last_voted_fork_slots_aggregate: Some(LastVotedForkSlotsAggregateRecord {
+                    received: expected_messages
+                }),
+            }
+        )
+    }
+
+    fn change_proto_file_readonly(wen_restart_proto_path: &PathBuf, readonly: bool) {
+        let mut perms = std::fs::metadata(wen_restart_proto_path)
+            .unwrap()
+            .permissions();
+        perms.set_readonly(readonly);
+        std::fs::set_permissions(wen_restart_proto_path, perms).unwrap();
+    }
+
+    #[test]
+    fn test_wen_restart_initialize_failures() {
+        solana_logger::setup();
+        let ledger_path = get_tmp_ledger_path_auto_delete!();
+        let test_state = wen_restart_test_init(&ledger_path);
+        let last_vote_bankhash = Hash::new_unique();
+        let mut last_voted_fork_slots = test_state.last_voted_fork_slots.clone();
+        last_voted_fork_slots.reverse();
+        let mut file = File::create(&test_state.wen_restart_proto_path).unwrap();
+        file.write_all(b"garbage").unwrap();
+        assert!(initialize(
+            &test_state.wen_restart_proto_path,
+            VoteTransaction::from(Vote::new(last_voted_fork_slots.clone(), last_vote_bankhash)),
+            test_state.blockstore.clone()
+        )
+        .err()
+        .unwrap()
+        .to_string()
```

Comment:
Can we remove this `.err().unwrap().to_string()` pattern from all the tests? We can just do an 

`assert_eq!(initialize(), Err())`

Reply:
Unfortunately you can't directly use assert_eq or even assert_match on anyhow::Error:
https://users.rust-lang.org/t/issues-in-asserting-result/61198/5

I have to use a unwrap_error().downcast().unwrap(), would you prefer that instead?

Reply:
wow so ugly 🤮 

yeah, I think let's downcast and match on the actual error.

Reply:
Done.